hobbify/habits/views/habits.py

```diff
@@ -0,0 +1,209 @@
+        if request.method == 'PATCH':
+            instance.done = request.data.done
+        return Response(instance, status=status.HTTP_200_OK)
+
+    @action(methods=['get'], detail=True)
+    def recurrences(self, request, *args, **kwargs):
+        """ Return the past recurrences of an habit """
+        recurrences = RecurrentInstance.objects.filter(habit=self.kwargs['id'], is_hidden=False)
+        data = [RecurrenceModelSerializer(r).data for r in recurrences]
+        return Response(data, status=status.HTTP_200_OK)
+
+    @action(detail=True, methods=['post'])
+    def recurrence(self, request, *args, **kwargs):
+        """ Create today's instance of an habit """
+        habit_id = self.kwargs['id']
+        habit = Habit.objects.filter(id=habit_id, is_hidden=False).first()
+        rule = self.set_rule(habit_id)
+        today = datetime.datetime.now().strftime("%Y-%m-%d")
+        all_recurrences = RecurrentInstance.objects.filter(habit=habit, is_hidden=False)
+
+        for s in all_recurrences:
+            date = str(s.recurrence).partition(' ')[0]
+            if date == today:
+                return Response(
+                    "Recurrence alredy exists in database, please retrieve it.",
+                    status=status.HTTP_406_NOT_ACCEPTABLE
+                )
+
+        all_instances = [{"date": s} for s in rule]
+        todays_intance = ""
+
+        for obj in all_instances:
+            date = str(obj['date']).partition(' ')[0]
+            if today == date:
+                todays_intance = obj['date']
+                break
+
+        if todays_intance == "":
+            return Response(
+                "Cannot create an instance if it does not match its recurrent rule",
+                status=status.HTTP_406_NOT_ACCEPTABLE
+            )
+
+        instance = RecurrentInstance.objects.create(
+            habit=habit,
+            done=False,
+            rule=str(rule),
+            recurrence=str(todays_intance)
+        )
+        data = RecurrenceModelSerializer(instance).data
+
+        return Response(
+            data,
+            status=status.HTTP_201_CREATED
+        )
+
+    @action(detail=True, methods=['get'])
+    def today(self, request, *args, **kwargs):
+        """ Return today's instance of an habit """
+        habit_id = self.kwargs['id']
+        rule = self.set_rule(habit_id)
+        today = datetime.datetime.now().strftime("%Y-%m-%d")
+
+        instances = RecurrentInstance.objects.filter(is_hidden=False, habit=habit_id)
+        instances_data = [RecurrenceModelSerializer(instance).data for instance in instances]
+        todays_intance = {}
+
+        for info in instances_data:
+            instance_date = str(info['recurrence']).partition(' ')[0]
+            print(instance_date, today)
+            if instance_date == today and str(info['rule']) == str(rule):
+                todays_intance = info
```

Comment:
This is highly inefficient. You will be searching through the whole list of instances.